elastic/kibana
# [ContentManagement] Inspector flyout

This PR adds a new `@kbn/content-management-inspector` package be able to inspect a content (saved object)
from anywhere in Kibana. It is has been integrated in the `<TableListView />` (`@kbn/content-management-table-list` package).

**Note:** The inspector has been integrated in the table list view but has _not_ been enabled anywhere yet (dashboard, visualization, maps, graph). Integration will come in a separate PR.

## How to test

To be able to test the inspector we need to enable it in one of the consuming apps.

* Open the `src/plugins/dashboard/public/application/listing/dashboard_listing.tsx` file
* Add an `inspector` prop to the `<TableListView ...>`

```ts
<TableListView<DashboardSavedObjectUserContent>
  createItem={!showWriteControls ? undefined : createItem}
  ...
  // Add this
  inspector={{
    isReadonly: false,
    onSave: async (args) => {
      console.log('Received', args);
    },
  }}
>
```

* Test the metadata form. You should receive the updated data back in the `onSave()` handler
* Set the `isReadonly` to `false` and make sure that none of the attributes can be edited

### `@kbn/content-management-inspector` (new package)

The package exposes:

* `InspectorProvider`: A React context provider
* `InspectorKibanaProvider`: A React context provider to wire Kibana dependencies
* `useOpenInspector`: A hook to open the inspector

### Usage

```ts
import { useOpenInspector } from '@kbn/content-management-inspector';

export const MyComponent = () => {
  const openInspector = useOpenInspector();

  const onClick = () => {
    openInspector({
      item: {
        id: 'foo-123', // required (string)
        title: 'My item', // required (string)
        description: '', // optional (string)
        tags: [], // optional (Array of Saved object references)
      },
      isReadonly: false, // optional (defaults to "true"),
      entityName: 'dashboard', // the item entity name
      // onSave handler is only required when "isReadonly" is set to "false"
      onSave: async ({ id:, title, description, tags }) => {
        // Do something with the updated data
      }
    })
  }

  return <EuiButton onClick{onClick}>Open inspector<EuiButton>
}
```

### `@kbn/content-management-table-list`

The package has been updated to integrate the inspector. There is a new `inspector` prop that can be passed to the `<TableListView />` component to control the inspector.
If it is not defined, the inspector is disabled.

### Usage

```ts
<TableListView
  ...
  inspector{{
    isReadonly: false,
    onSave: async ({ id:, title, description, tags }) => {
      // Update the saved object
    }
  }}
/>
```

## Screenshots

<img width="1693" alt="Screenshot 2022-11-02 at 10 34 21" src="https://user-images.githubusercontent.com/2854616/199522497-c121a0bd-61fc-4758-ac18-5889906f20e3.png">

**Readonly**

<img width="1696" alt="Screenshot 2022-11-02 at 10 32 56" src="https://user-images.githubusercontent.com/2854616/199522546-f39fe7b2-9f18-46e5-8bcc-8f5057832c89.png">

**Edit capability**

<img width="1698" alt="Screenshot 2022-11-02 at 10 33 26" src="https://user-images.githubusercontent.com/2854616/199522582-245ea054-dd37-4a44-80cb-03b3a0fb7707.png">



Fixes https://github.com/elastic/kibana/issues/143939
